tfinstall/tfinstall.go

```diff
@@ -0,0 +1,346 @@
+VjC+8YlRhS2HNI/8TAUCXn0BIQIbAwULCQgHAgYVCgkICwIEFgIDAQIeAQIXgAAK
+CRBRhS2HNI/8TJITCACT2Zu2l8Jo/YLQMs+iYsC3gn5qJE/qf60VWpOnP0LG24rj
+k3j4ET5P2ow/o9lQNCM/fJrEB2CwhnlvbrLbNBbt2e35QVWvvxwFZwVcoBQXTXdT
++G2cKS2Snc0bhNF7jcPX1zau8gxLurxQBaRdoL38XQ41aKfdOjEico4ZxQYSrOoC
+RbF6FODXj+ZL8CzJFa2Sd0rHAROHoF7WhKOvTrg1u8JvHrSgvLYGBHQZUV23cmXH
+yvzITl5jFzORf9TUdSv8tnuAnNsOV4vOA6lj61Z3/0Vgor+ZByfiznonPHQtKYtY
+kac1M/Dq2xZYiSf0tDFywgUDIF/IyS348wKmnDGjuQENBFMORM0BCADWj1GNOP4O
+wJmJDjI2gmeok6fYQeUbI/+Hnv5Z/cAK80Tvft3noy1oedxaDdazvrLu7YlyQOWA
+M1curbqJa6ozPAwc7T8XSwWxIuFfo9rStHQE3QUARxIdziQKTtlAbXI2mQU99c6x
+vSueQ/gq3ICFRBwCmPAm+JCwZG+cDLJJ/g6wEilNATSFdakbMX4lHUB2X0qradNO
+J66pdZWxTCxRLomPBWa5JEPanbosaJk0+n9+P6ImPiWpt8wiu0Qzfzo7loXiDxo/
+0G8fSbjYsIF+skY+zhNbY1MenfIPctB9X5iyW291mWW7rhhZyuqqxN2xnmPPgFmi
+QGd+8KVodadHABEBAAGJATwEGAECACYCGwwWIQSRpuf4XQXGVjC+8YlRhS2HNI/8
+TAUCXn0BRAUJEvOKdwAKCRBRhS2HNI/8TEzUB/9pEHVwtTxL8+VRq559Q0tPOIOb
+h3b+GroZRQGq/tcQDVbYOO6cyRMR9IohVJk0b9wnnUHoZpoA4H79UUfIB4sZngma
+enL/9magP1uAHxPxEa5i/yYqR0MYfz4+PGdvqyj91NrkZm3WIpwzqW/KZp8YnD77
+VzGVodT8xqAoHW+bHiza9Jmm9Rkf5/0i0JY7GXoJgk4QBG/Fcp0OR5NUWxN3PEM0
+dpeiU4GI5wOz5RAIOvSv7u1h0ZxMnJG4B4MKniIAr4yD7WYYZh/VxEPeiS/E1CVx
+qHV5VVCoEIoYVHIuFIyFu1lIcei53VD6V690rmn0bp4A5hs+kErhThvkok3c
+=+mCN
+-----END PGP PUBLIC KEY BLOCK-----`
+
+type findOption struct {
+	find func() (string, error)
+	name string
+}
+
+// each represents a strategy
+// strategies are executed in order
+// each strategy can return a string, or not
+
+func ExactPath(execPath string) *findOption {
+	opt := &findOption{
+		name: "ExactPath",
+	}
+	opt.find = func() (string, error) {
+		return execPath, nil
+	}
+
+	return opt
+}
+
+func LookPath() *findOption {
+	opt := &findOption{
+		name: "LookPath",
+	}
+	opt.find = func() (string, error) {
+		p, err := exec.LookPath("terraform")
+		if err != nil {
+			return "", err
+		}
+		return p, nil
+	}
+
+	return opt
+}
+
+func LatestVersion(installDir string) *findOption {
+	opt := &findOption{
+		name: "LatestVersion",
+	}
+	opt.find = func() (string, error) {
+		v, err := latestVersion()
+		if err != nil {
+			return "", err
+		}
+
+		return downloadWithVerification(v, installDir)
+	}
+
+	return opt
+}
+
+func ExactVersion(tfVersion string, installDir string) *findOption {
+	opt := &findOption{
+		name: "ExactVersion",
+	}
+	opt.find = func() (string, error) {
+		// validate version
+		_, err := version.NewVersion(tfVersion)
+		if err != nil {
+			return "", err
+		}
+
+		return downloadWithVerification(tfVersion, installDir)
+	}
+
+	return opt
+}
+
+func Find(opts ...*findOption) (string, error) {
+	var terraformPath string
+	var errors []string
+
+	// go through the options in order
+	// until a valid terraform executable is found
+	for _, opt := range opts {
+		p, err := opt.find()
+		if err != nil {
+			// TODO KEM make logging configurable
+			findErr := fmt.Sprintf("did not locate terraform executable using strategy %s (error: %s)", opt.name, err)
+			errors = append(errors, findErr)
```

Comment:
Would it make sense to use something like go-multierror here?

Reply:
Why don't we just stop at the first error, can the methods just return empty string and nil if not found?